Show all errors, not just the first, when compiling TEAL.

jannotti opened the pull request:
## Summary
Previously, we used the usual Go convention of returning an error to
the caller throughout assembly, as soon as they occurred.  This meant
that assembly failed, as a whole, on the first bad line.  This model
also didn't work well for warnings, because they did not have the
context to do good reporting (they did not know the filename, for
example), but had no way to pass the warning up to the caller who did
have the info (as errors did).  So this change also allows warnings to
report the file and line, for use by external tools.

The change is to accumulate errors and warnings in the OpStream
structure, rather than return immediately (or report immediately, in
the case of warnings).  After assembly the warnings and errors are
reported to stderr, and a summary error is used in Go to signal the
caller that assembly failed.

This allows external tools to use `goal clerk compile` to show all of
a script's problems at once.  I've added support to emacs in
teal-mode. We should consider a vscode plugin.


## Test Plan

Updated and ran unit tests. (And added one for testing that multiple errors are returned.)

brianolson reviewed (commented):
This definitely makes the TEAL assembler more like other good compiler tools. One more hopefully little thing to add: don't runaway error spew, stop after 10 or so.

jannotti commented on data/transactions/logic/assembler.go:
```diff
@@ -1062,62 +1067,68 @@ func (ops *OpStream) assemble(fin io.Reader) error {
-			if err != nil {
-				return err
-			}
+			asmFunc(ops, &spec, fields[1:])
```
Good point, will do.  I was thinking TEAL wouldn't be susceptible to the same concerns of a single error causing a cascading catastrophe because it's line-oriented.  But it's still probably the right thing to do.

jannotti replied:
Since the assembly isn't going to be slow enough to matter, I decide the code will be more readable if I just proceed as now, but only report up to 10 errors.  So if I get cascading errors, the user experience will still be ok, but I don't need to insert checks at each error location to decide if I should quit now.

You comment made me revisit a number of places where I was returning early during assembling a line.  I found that I could give much better errors for future lines by "correcting" the error and continuing.  That way the internal type stack is maintained, even though you have a typo in your code, for example.  So thanks for that.

jannotti commented on data/transactions/logic/assembler.go:
```diff
@@ -1062,62 +1067,68 @@ func (ops *OpStream) assemble(fin io.Reader) error {
+	saved := ops.sourceLine
+	raw := ops.pending.Bytes()
 	for _, lr := range ops.labelReferences {
+		ops.sourceLine = lr.sourceLine
```
I decided to address this more directly by tracking the undefined references that have already been reported, and avoid repeats.  (Though if you have a great many different undefined references, that will be handled by the general case of stopping error reports after 10.)

algorandskiy reviewed (commented):
Good work, I like the idea. But I don't really want to encourage OpStream usage outside of logic package. Please consider alternative design if it make sense.

jannotti commented on data/transactions/logic/assembler.go:
```diff
@@ -1155,81 +1211,99 @@ func (ops *OpStream) Bytes() (program []byte, err error) {
-	program, _, err := AssembleStringWithVersionEx(text, version)
-	return program, err
+// AssembleString takes an entire program in a string and assembles it to bytecode using AssemblerDefaultVersion
+func AssembleString(text string) (*OpStream, error) {
```
What third party tools do we need to think about?  Is there anyone linking with our modules?

As it was, the interface was already returning three things - the program bytes, the offset map, and err.  That was already a bit odd - the offset map was only used in tests.  Adding teal errors and warnings to the return signature would have made the function return five values which seems extremely unwieldy.  By combining the related ideas into a single abstraction, the OpStream, we keep related things together.

jannotti replied:
I'm willing the rename the OpStream to "Assembly" if that makes sense.  I think it would be a better name for what is now.  It might even bring in the PragmaStream functionality.

algorandskiy replied:
>What third party tools do we need to think about? Is there anyone linking with our modules?

I'm not aware of any but who knows.

> Adding teal errors and warnings to the return signature would have made the function return five values which seems extremely unwieldy.

Not adding things to return but package everything into err. Agree, this does not work for warnings.

jannotti commented on data/transactions/logic/fields.go:
```diff
@@ -230,7 +230,7 @@ var TxnTypeNames = []string{
 }
 
 // map TxnTypeName to its enum index, for `txn TypeEnum`
-var txnTypeIndexes map[string]int
```
There are several of these maps whose only purpose is to extract values for insertion into a TEAL program.  TEAL programs only operate on uint64s.  So some of these maps were already set up to map to uint64 so they could be indexed and the value used directly, while others were like this, and the value needed to be converted each time.  At the very least, I felt they should all be done the same way, and I went this direction because the only use was TEAL.  (Fundamentally, the only purpose for the map is to provide a name in TEAL, so TEAL is the only intended user, ever.)

jannotti commented:
Is this acceptable to merge?  I think I've addressed the concerns of @brianolson and maybe @algorandskiy. For Pavel, I didn't make the changes because the need to handle warnings by handing out a structure (the OpStream) make it less valuable to figure out a way to merge all of the errors into one - the interface had to change anyway, and I think a list of errors is conceptually cleaner than embedding them inside a kind of "MultiError" I would have done the MultiError thing if the interface would have otherwise been stable.  I also intend to add more warnings under some other issues soon - and they will be important security warnings, so I don't want to encourage an interface that doesn't return them.

There was also the question of array indexes as uint64.  I agree it seems weird, but holding these maps from names to indexes is only intended for TEAL constants, which are uint64.
